Add Dutch translation to How Does React Tell a Class From a Function (#471)
gaearon/overreacted.io: src/pages/how-does-react-tell-a-class-from-a-function/index.nl.md

```jsx{5}
// Als Greeting een functie is
const result = Greeting(props); // <p>Hello</p>

// Als Greeting een class is
const instance = new Greeting(props); // Greeting {}
const result = instance.render(); // <p>Hello</p>
```

Laten we in een notendop kijken wat de `new` operator doet in JavaScript.

---

Zelfs toen JavaScript vroeger geen classes had kon je een vergelijkbaar patroon als die van een class gebruiken door functies in te zetten. **Heel concreet: je kan iedere functie dezelfde rol geven als een class constructor door `new` te plaatsen voor het aanroepen:**

```jsx
// Een normale functie
function Person(name) {
  this.name = name;
}

var fred = new Person('Fred'); // ✅ Person {name: 'Fred'}
var george = Person('George'); // 🔴 Werkt niet
```

Zelfs nu dat JavaScript classes heeft werkt bovenstaande code nog steeds! Probeer het maar eens in DevTools.

Als je `Person(‘Fred’)` aanroept **zonder** `new`, verwijst `this` naar iets globaals en onhandigs (bijvoorbeeld `window` of `undefined`). Onze code zou dus kunnen crashen of iets raars doen zoals `window.name` creëren.

> Door `new` te gebruiken zeggen we eigenlijk: “Hey JavaScript, ik weet dat `Person` gewoon een functie is. Maar laten deze gebruiken alsof het een class constructor is. **Maak een `{}` object en verwijs `this` binnen de `Person` functie naar dat `{}` object zodat ik dingen zoals `this.name` kan toewijzen. Laat me daarna dat object weer zien.”**

Reviewer: ```suggestion
Door `new` te gebruiken zeggen we eigenlijk: “Hey JavaScript, ik weet dat `Person` gewoon een functie is. Maar laten deze gebruiken alsof het een class constructor is. **Maak een `{}` object en verwijs `this` binnen de `Person` functie naar dat `{}` object zodat ik dingen zoals `this.name` kan toewijzen. Geef mij daarna dat object weer terug.”**
```

```jsx
class Person {
  constructor(name) {
    this.name = name;
  }
  sayHi() {
    alert('Hi, I am ' + this.name);
  }
}

let fred = new Person('Fred');
fred.sayHi();
```

*Die duidelijkheid over wat de developer ergens mee wil bereiken is enorm belangrijk.* Het is dus belangrijk om hier rekening mee te houden bij zowel het ontwerpen van programmeertalen als API’s.

Als je een functie schrijft kan JavaScript niet zelf inschatten of deze moet worden aangeroepen met `alert()` of dat het een constructor moet zijn zoals `new Person()`. `new` vergeten bij het aanroepen van `Person` kan voor verwarrende resultaten zorgen.

**Dankzij de syntax van class is het mogelijk te zeggen: “Dit is niet zomaar een functie - het is een class en het heeft een constructor”.** Als je `new` vergeet te gebruiken wanneer je deze aanroept geeft JavaScript een error:

```jsx
let fred = new Person('Fred');
// ✅ Als Person een functie is: werkt prima
// ✅ Als Person een class is: werkt ook prima

let george = Person('George'); // We zijn ‘new’ vergeten

> // 😳 Als Person een constructor-achtige functie is: verwarrend geïmplementeerd.

Reviewer: ```suggestion
// 😳 Als Person een constructor-achtige functie is: verwarrend gedrag
```

Dit betekent echter wel dat React `new` moet gebruiken voordat een class aangeroepen wordt. Een class kan niet aangeroepen worden als een normale functie, sinds JavaScript dit als een error zou behandelen.

```jsx
class Counter extends React.Component {
  render() {
    return <p>Hello</p>;
  }
}

// 🔴 Dit is in React niet zomaar mogelijk:
const instance = Counter(props);
```

Dit vraagt om problemen.

---

Voordat we gaan kijken hoe React dit oplost is het belangrijk om te realiseren dat de meeste gebruikers van React compilers zoals Babel gebruiken om moderne functionaliteiten zoals classes te kunnen gebruiken in oudere browsers. Dus we moeten rekening houden met compilers in het ontwerp van React.

In oudere versies van Babel konden classes aangeroepen worden zonder `new`. Dit is echter opgelost  - door middel van wat extra code.

```jsx
function Person(name) {
  // Een versimpelde versie van de Babel output:
  if (!(this instanceof Person)) {
    throw new TypeError("Cannot call a class as a function");
  }
  // Onze code:
  this.name = name;
}

new Person('Fred'); // ✅ Okay

> Person('George'); // 🔴 Een class kan je niet aanroepen als een functie.

Reviewer: Should we translate this line? I'm asking since line 193 where the error is defined hasn't been translated either.

Author: Agreed!

Dus, kan React gewoon checken of iets een class is of niet?

Nou, zo makkelijk is dat dus niet. [Zelfs als we een class van een functie zouden kunnen onderscheiden in JavaScript](https://stackoverflow.com/questions/29093396/how-do-you-check-the-difference-between-an-ecmascript-6-class-and-function) zou dit niet werken voor classes die zijn verwerkt door tools zoals Babel. Voor de browser is het namelijk gewoon een functie. Pech voor React.

---

Oke, dus misschien zou React gewoon `new` kunnen gebruiken op iedere call? Nou… jammer genoeg zou ook dat niet echt werken.

Normale functies die worden aangeroepen met `new` krijgen een object instance zoals `this`. Dat is op zich wenselijk voor functies die als constructor zijn geschreven (zoals `Person` hierboven) maar het zou verwarrend zijn voor function components:

```jsx
function Greeting() {
  // We verwachten niet dat `this` hier ook maar enigszins een vorm van een instance is.
  return <p>Hello</p>;
}
```

Dit hoeft natuurlijk geen probleem te zijn. Toch zijn er nog twee *andere* redenen waarom het juist wel een probleem is.

---

> De eerste reden waarom `new` gebruiken niet altijd zou werken is dat native arrow functies aanroepen met `new` een error geeft (behalve als ze zijn gecompiled door Babel):

Reviewer: ```suggestion
De eerste reden waarom `new` gebruiken niet altijd zou werken is dat native arrow functies aanroepen met `new` een error geeft (behalve als ze zijn gecompileerd door Babel):
```

```jsx
(() => {}).prototype // undefined
(function() {}).prototype // {constructor: f}
```

Maar dit [zou niet werken](https://github.com/facebook/react/issues/4599#issuecomment-136562930) voor functies die al zijn compiled door Babel. Misschien niet echt een big deal, maar er is ook nog een andere reden die ervoor zorgt dat deze aanpak niet een goed idee is.

---

Een andere reden waarom we niet altijd `new` kunnen gebruiken is omdat het React ervan zou weerhouden om components te ondersteunen die strings of andere primitieve types teruggeven.

```jsx
function Greeting() {
  return 'Hello';
}

Greeting(); // ✅ 'Hello'
new Greeting(); // 😳 Greeting {}
```

Dit heeft wederom te maken met de gekkigheden van het ontwerp van de [`new` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new). Zoals we eerder zagen vertelt `new` de JavaScript engine om een object te maken, deze om te zetten in `this` binnen de functie en dit object later terug te geven als resultaat van `new`.

> Echter, JavaScript staat het functies die zijn aangeroepen met `new` ook toe om de return value van `new` te overschrijven door een ander object terug te geven. Waarschijnlijk omdat dit handig zou zijn voor patterns zoals pooling waarbij we instances willen hergebruiken:

Reviewer: ```suggestion
Echter, JavaScript staat het functies die zijn aangeroepen met `new` ook toe om de return value van `new` te overschrijven door een ander object terug te geven. Waarschijnlijk omdat dit handig zou zijn voor patterns zoals pooling waarbij we instanties willen hergebruiken:
```

Not sure if we should translate it or keep it in English. We do have a good Dutch word for "instances", but it might be less clear to the reader. The same goes for "gecompiled" and "gecompileerd".

Dat is onacceptabel dus we moeten een tussenweg zien te vinden.

---

Wat hebben we tot nu toe geleerd? React *moet* classes (inclusief Babel output) aanroepen met `new` maar het moet normale functies of arrow functies (inclusief Babel output) aanroepen *zonder* `new`. En er is geen betrouwbare manier om deze twee van elkaar te onderscheiden.

**Als we een algemeen probleem niet kunnen oplossen, kunnen we dan misschien wel een meer specifiek probleem oplossen?**

Als je een component als een class definieert wil je waarschijnlijk `React.Component` gebruiken voor ingebouwde methoden zoals `this.setState()`. **In plaats van alle classes proberen te detecteren, kunnen we ook gewoon op zoek gaan naar `React.Component` afstammelingen?**

Spoiler: dit is precies wat React doet.

---

Misschien is de meest idiomatische manier om te checken of `Greeting` een React component class is door dit te testen: `Greeting.prototype instanceof React.Component`:

```jsx
class A {}
class B extends A {}

console.log(B.prototype instanceof A); // true
```

Ik weet wat je nu denkt. Wat is hier zojuist gebeurt?! Om dit uit te leggen moeten we JavaScript prototypes begrijpen.

> Misschien ben je bekend met de ‘prototype chain’. Ieder object in JavaScript kan een ‘prototype’ hebben. Wanneer we `fred.sayHi()` schrijven maar `fred` heeft geen `sayHi` property, kijken we naar de `sayHi` property op `fred`’s prototype. Als we deze hier niet kunnen vinden kijken we naar de volgende prototype in de schakel - `fred`’s prototype’s prototype. Ga zo maar door.

Reviewer: ```suggestion
Misschien ben je bekend met de ‘prototype chain’. Ieder object in JavaScript kan een ‘prototype’ hebben. Wanneer we `fred.sayHi()` schrijven maar `fred` heeft geen `sayHi` property, kijken we naar de `sayHi` property op `fred`’s prototype. Als we deze hier niet kunnen vinden kijken we naar de volgende prototype in de schakel - de `prototype` van `fred`’s `prototype`. Ga zo maar door.
```

I think this reads a little better.

  alert('Hi, I am ' + this.name);
}

var fred = new Person('Fred'); // Zet ‘fred.__proto__’ naar ‘Person.prototype’
```

En die `__proto__` chain is hoe JavaScript properties opzoekt:

```jsx
fred.sayHi();
// 1. Heeft fred een sayHi property? Nee.
// 2. Heeft fred.__proto__ een sayHi property? Ja! Roep maar aan!

fred.toString();
// 1. Heeft fred een toString property? Nee.
// 2. Heeft fred.__proto__ een toString property? Nee.
// 3. Heeft fred.__proto__.__proto__ een toString property? Ja! Roep maar aan!
```

In de praktijk zou je eigenlijk nooit `__proto__` hoeven aanraken in je code behalve als je iets aan het debuggen bent dat te maken heeft met de prototype chain. Als je iets beschikbaar wil maken op `fred.__proto__` moet je deze eigenlijk zetten op `Person.prototype`. Althans, dat is hoe het ontwerp in eerste instantie was.

De `__proto__` property was niet eens bedoelt om beschikbaar gemaakt te worden door browsers omdat de prototype chain werd gezien als een intern concept. Maar sommige browsers hebben `__proto__` toegevoegd en uiteindelijk werd het heel erg gestandaardiseerd. (wel deprecated omdat er een preferentie kwam voor `Object.getPrototypeOf()`).

> **Toch blijf ik het verwarrend vinden dat een property die `prototype` genoemd is, niet de prototype teruggeeft van een value. (bijvoorbeeld, `fred.prototype` is `undefined` omdat `fred` geen functie is). Persoonlijk denk ik dat dit een van de grootste redenen is dat zelfs developers met veel ervaring moeite hebben met het begrijpen van JavaScript prototypes.

Reviewer: ```suggestion
**Toch blijf ik het verwarrend vinden dat een property die `prototype` genoemd is, niet de prototype teruggeeft van een value** (bijvoorbeeld, `fred.prototype` is `undefined` omdat `fred` geen functie is). Persoonlijk denk ik dat dit een van de grootste redenen is dat zelfs developers met veel ervaring moeite hebben met het begrijpen van JavaScript prototypes.
```

---

> Dit is een lange post he? Ik zou zeggen dat we er voor ongeveer 80% zijn. Houd vol.

Reviewer: ```suggestion
Dit is een lange post hè? Ik zou zeggen dat we er voor ongeveer 80% zijn. Houd vol.
```

Author: I will add a different solution!
